lib/LanguagePicker.js

```diff
@@ -19,50 +42,86 @@ function LanguagePicker(lang = 'en', config = {}) {
-  if (fallbacks.indexOf('en') === -1) {
-    fallbacks.push('en');
-  }
-
-  // Fallback to nameTag
-  if (this.nameTag) {
-    fallbacks.push(this.nameTag);
-  }
+  // Remove duplicates
+  fallbacks = fallbacks.filter((item, i) => fallbacks.indexOf(item) === i);
 
   // Add prefix to all languages if exists
   // eslint-disable-next-line arrow-body-style
   fallbacks = fallbacks.map((code) => {
     return code === this.nameTag ? code : this.prefix + code;
   });
 
+  // Store initial fallbacks
   this.fallbacks = fallbacks;
 }
 
 LanguagePicker.prototype.newProcessor = function newProcessor() {
-  const { fallbacks, nameTag, prefix } = this;
-  let bestChoiceLang = nameTag || (prefix ? `${prefix}en` : 'en');
-  let bestChoiceValue;
   let firstChoiceValue;
+  const prefixedEnglish = this.prefix ? `${this.prefix}en` : 'en';
+  const prefixedLangScript = `-${this.langScript}`;
 
   return {
-    addValue(lang, value) {
-      const langIndex = fallbacks.indexOf(lang);
-      if (
-        langIndex > -1 &&
-        langIndex <= fallbacks.indexOf(bestChoiceLang)
-      ) {
-        // Given language is a better fallback
-        bestChoiceLang = lang;
-        bestChoiceValue = value;
+    addValue: (lang, value) => {
+      this.values[lang] = value;
+
+      if (!firstChoiceValue) {
+        firstChoiceValue = value;
+      }
+    },
+    getResult: () => {
+      let val;
+
+      // Get the best value from the best language fallback:
+      // 1. Requested language
+      val = this.values[this.userLang];
+      if (val) {
+        return val;
+      }
+
+      // 2. Fallback language from fallbacks.json
+      for (const fallback of this.fallbacks) {
+        val = this.values[fallback];
+        if (val) {
+          return val;
+        }
+      }
+
+      // 3. Any language with suffix of same script as requested language
+      const valueLangCodes = Object.keys(this.values);
+      for (const langCode of valueLangCodes) {
+        if (langCode.endsWith(prefixedLangScript)) {
+          return this.values[langCode];
+        }
+      }
+
+      // 4. Another language with the same script
+      for (const langCode of this.languagesInScript) {
+        val = this.values[langCode];
+        if (val) {
+          return val;
+        }
+      }
+
+      // 5. English
+      val = this.values[prefixedEnglish];
+      if (val) {
+        return val;
+      }
+
+      // 6. Any language with suffix 'Latn'
+      // Only do this if the script isn't already latin;
+      // otherwise, we've done it in step 3
+      for (const langCode of valueLangCodes) {
+        if (langCode.endsWith('-Latn')) {
+          return this.values[langCode];
+        }
       }
 
-      // Store the value of the first value given
-      firstChoiceValue = firstChoiceValue !== undefined ?
-        firstChoiceValue : value;
+      // If nothing is found:
+      // - If there's a name tag, return it
+      // - If there's no name tag, return first choice
```

Comment:
sorry, a tiny nitpick - could you change the comment to " Otherwise, return first choice". I keep misreading it because it only differs in two letters with previous line.

Reply:
Yup, fair. Changing.